packages/go_router/lib/src/route.dart

```diff
@@ -1241,25 +1285,71 @@ class StatefulNavigationShellState extends State<StatefulNavigationShell>
   }
 
   void _updateCurrentBranchStateFromWidget() {
+    _preloadBranches();
+
     final StatefulShellBranch branch = route.branches[widget.currentIndex];
     final ShellRouteContext shellRouteContext = widget.shellRouteContext;
     final RouteMatchList currentBranchLocation =
         _scopedMatchList(shellRouteContext.routeMatchList);
 
-    final _RestorableRouteMatchList branchLocation =
-        _branchLocation(branch, false);
-    final RouteMatchList previousBranchLocation = branchLocation.value;
-    branchLocation.value = currentBranchLocation;
-    final bool hasExistingNavigator =
-        _branchNavigators[branch.navigatorKey] != null;
+    final _StatefulShellBranchState branchState =
+        _branchStateFor(branch, false);
+    final RouteMatchList previousBranchLocation = branchState.location.value;
+    branchState.location.value = currentBranchLocation;
+    final bool hasExistingNavigator = branchState.navigator != null;
 
     /// Only update the Navigator of the route match list has changed
     final bool locationChanged =
         previousBranchLocation != currentBranchLocation;
     if (locationChanged || !hasExistingNavigator) {
-      _branchNavigators[branch.navigatorKey] = shellRouteContext
-          .navigatorBuilder(branch.observers, branch.restorationScopeId);
+      branchState.navigator = shellRouteContext._buildNavigatorForCurrentRoute(
+          branch.observers, branch.restorationScopeId);
     }
+
+    _cleanUpObsoleteBranches();
+  }
+
+  void _preloadBranches() {
+    for (int i = 0; i < route.branches.length; i++) {
+      final StatefulShellBranch branch = route.branches[i];
+      if (i != currentIndex && branch.preload && !_isBranchLoaded(branch)) {
+        // Find the match for the current StatefulShellRoute in matchList
+        // returned by _effectiveInitialBranchLocation (the initial location
+        // should already have been validated by RouteConfiguration).
+        final RouteMatchList matchList = _router.configuration
+            .findMatch(widget._effectiveInitialBranchLocation(i));
+        ShellRouteMatch? match;
+        matchList.visitRouteMatches((RouteMatchBase e) {
+          match = e is ShellRouteMatch && e.route == route ? e : match;
+          return match == null;
+        });
+        assert(match != null);
+
+        final Widget navigator = widget.shellRouteContext.navigatorBuilder(
+          branch.navigatorKey,
+          match!,
```

Comment:
`!` can be removed on this line

Reply:
Oh? Maybe I'm missing something, but that gives me an error (since match is ShellRouteMatch?). Would have been nice if the assert would have that effect though.